File: C++/Reverse_Linked_List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        ListNode *newHead = NULL;
        reverseListHelper(head, &newHead);
        return newHead;
    }
    ListNode* reverseListHelper(ListNode* head, ListNode **newHead) {
        if (head == NULL) return NULL;
        if (head->next == NULL) {
            *newHead = head;
        }
        else {
            reverseListHelper(head->next, newHead)->next = head;
            head->next = NULL;
        }
        return head;
    }
};

File: Go/Reverse_LinkedList.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reverseListRec(head *ListNode, newHead **ListNode) *ListNode{
    if (head == nil || head.Next == nil) {
        *newHead = head;
    } else {
        reverseListRec(head.Next, newHead).Next = head
        head.Next = nil
    }
    return head
}
func reverseList(head *ListNode) *ListNode {
    ret := &ListNode{}
    reverseListRec(head, &ret)
    return ret
}
